pkg/resources/data.go

```diff
@@ -689,13 +690,19 @@ func GetCSIMigrationFeatureGates(cluster *kubermaticv1.Cluster) []string {
 		// The CSIMigrationNeededAnnotation is removed when all kubelets have
 		// been migrated.
 		if kubermaticv1helper.ClusterConditionHasStatus(cluster, kubermaticv1.ClusterConditionCSIKubeletMigrationCompleted, corev1.ConditionTrue) {
-			// TODO: This feature flag was removed in k8s 1.22
 			if cluster.Spec.Cloud.Openstack != nil {
-				featureFlags = append(featureFlags, "CSIMigrationOpenStackComplete=true")
+				if lessThan21, _ := semver.NewConstraint("< 1.21.0"); lessThan21.Check(cluster.Spec.Version.Version) {
+					featureFlags = append(featureFlags, "CSIMigrationOpenStackComplete=true")
+				} else {
+					featureFlags = append(featureFlags, "InTreePluginOpenStackUnregister=true")
+				}
 			}
-			// TODO: This feature flag was removed in k8s 1.22
 			if cluster.Spec.Cloud.VSphere != nil {
-				featureFlags = append(featureFlags, "CSIMigrationvSphereComplete=true")
+				if lessThan22, _ := semver.NewConstraint("< 1.22.0"); lessThan22.Check(cluster.Spec.Version.Version) {
```

Comment:
The `InTreePluginvSphereUnregister` feature gate is available since 1.21, so I think we should use it for 1.21 as well. `CSIMigrationvSphereComplete` is deprecated already since 1.21, but completely removed in 1.22.